internal/ctrlflow/trash.go

```diff
@@ -0,0 +1,513 @@
+	if types.AssignableTo(from, to) {
+		return expr
+	}
+
+	castExpr, err := t.typeConverter.Convert(to)
+	if err != nil {
+		panic(err)
+	}
+	return ah.CallExpr(&ast.ParenExpr{X: castExpr}, expr)
+}
+
+func (t *trashGenerator) chooseRandomVar(typ types.Type, vars map[string]*definedVar) ast.Expr {
+	var candidates []string
+	for name, d := range vars {
+		if canConvert(d.Type, typ) {
+			candidates = append(candidates, name)
+		}
+	}
+	if len(candidates) == 0 {
+		return nil
+	}
+
+	targetVarName := candidates[t.rand.Intn(len(candidates))]
+	targetVar := vars[targetVarName]
+	targetVar.AddRef()
+
+	return t.convertExpr(targetVar.Type, typ, ast.NewIdent(targetVarName))
+}
+
+func (t *trashGenerator) chooseRandomGlobal(typ types.Type) ast.Expr {
+	var candidates []*types.Var
+	for _, global := range t.globals {
+		if canConvert(global.Type(), typ) {
+			candidates = append(candidates, global)
+		}
+	}
+	if len(candidates) == 0 {
+		return nil
+	}
+
+	targetGlobal := candidates[t.rand.Intn(len(candidates))]
+
+	var globalExpr ast.Expr
+	if pkgIdent := t.importNameResolver(targetGlobal.Pkg()); pkgIdent != nil {
+		globalExpr = ah.SelectExpr(pkgIdent, ast.NewIdent(targetGlobal.Name()))
+	} else {
+		globalExpr = ast.NewIdent(targetGlobal.Name())
+	}
+	return t.convertExpr(targetGlobal.Type(), typ, globalExpr)
+}
+
+func (t *trashGenerator) generateRandomConst(p types.Type, rand *mathrand.Rand) ast.Expr {
+	var candidates []types.Type
+	for typ := range valueGenerators {
+		if canConvert(typ, p) {
+			candidates = append(candidates, typ)
+		}
+	}
+
+	if len(candidates) == 0 {
+		panic(fmt.Errorf("unsupported type: %v", p))
+	}
+
+	generatorType := candidates[rand.Intn(len(candidates))]
+	generator := valueGenerators[generatorType]
+	return t.convertExpr(generatorType, p, generator(rand, under(p)))
+}
+
+func (t *trashGenerator) generateRandomValue(typ types.Type, vars map[string]*definedVar) ast.Expr {
+	if t.rand.Float32() < varProb {
+		if expr := t.chooseRandomVar(typ, vars); expr != nil {
+			return expr
+		}
+	}
+	if t.rand.Float32() < globalProb {
+		if expr := t.chooseRandomGlobal(typ); expr != nil {
+			return expr
+		}
+	}
+	return t.generateRandomConst(typ, t.rand)
+}
+
+func (t *trashGenerator) cacheMethods(vars map[string]*definedVar) {
+	for _, d := range vars {
+		typ := deref(d.Type)
+		if _, ok := t.methodCache[typ]; ok {
+			continue
+		}
+
+		var methods []*types.Func
+		switch typ := typ.(type) {
+		case *types.Named:
+			for i := 0; i < typ.NumMethods(); i++ {
+				if m := typ.Method(i); token.IsExported(m.Name()) && isSupportedSig(m) {
+					methods = append(methods, m)
+					if len(methods) > limitFunctionCount {
+						break
+					}
+				}
+			}
+		case *types.Interface:
+			for i := 0; i < typ.NumMethods(); i++ {
+				if m := typ.Method(i); token.IsExported(m.Name()) && isSupportedSig(m) {
+					methods = append(methods, m)
+					if len(methods) > limitFunctionCount {
+						break
+					}
+				}
+			}
```

Comment:
This should be `case *types.Named, *types.Interface`, if that doesn't actually work in go the duplication is fine.

Reply:
```
# mvdan.cc/garble/internal/ctrlflow
internal\ctrlflow\trash.go:337:24: typ.NumMethods undefined (type types.Type has no field or method NumMethods)
internal\ctrlflow\trash.go:338:17: typ.Method undefined (type types.Type has no field or method Method)        
```

Unfortunately, they don't have a shared interface.

Reply:
https://github.com/golang/go/issues/65031

Reply:
You can declare the shared interface yourself and use it here. You can even do it inline if you just need one or two simple methods, like:

```
case interface{Foo() []bar}:
```

Reply:
```
switch typ := typ.(type) {
		case interface {
			NumMethods() int
			Method(i int) *types.Func
		}:
			for i := 0; i < typ.NumMethods(); i++ {
				if m := typ.Method(i); token.IsExported(m.Name()) && isSupportedSig(m) {
					methods = append(methods, m)
					if len(methods) > limitFunctionCount {
						break
					}
				}
			}
		}
```

Not bad, but gofmt wont let me write in 1 line

Reply:
define it in a separate line, e.g. `type methodSet interface {...`